WordPress/Classes/Services/CommentService.m

```diff
@@ -1119,6 +1119,44 @@ - (void)updateCommentAndSave:(Comment *)comment withRemoteComment:(RemoteComment
     [[ContextManager sharedInstance] saveContext:self.managedObjectContext];
 }
 
+
+-(Comment *)restoreHierarchicalComment:(RemoteComment *)remoteComment
+                                 depth:(NSInteger)depth
+                             hierarchy:(NSString *)hierarchy {
+    // Fetch the ReaderPost
+    NSError *error;
+    NSFetchRequest *fetchRequest = [[NSFetchRequest alloc] initWithEntityName:NSStringFromClass([ReaderPost class])];
+    fetchRequest.predicate = [NSPredicate predicateWithFormat:@"postID = %@", remoteComment.postID];
+    NSArray *results = [self.managedObjectContext executeFetchRequest:fetchRequest error:&error];
+    if (error) {
+        DDLogError(@"Error fetching post with id %@: %@", remoteComment.postID, error);
+        return nil;
+    }
+    
+    ReaderPost *post = [results firstObject];
+    if (!post) {
+        return nil;
+    }
+    
+    Comment *comment = [self findCommentWithID:remoteComment.commentID fromPost:post];
+    
+    // If the Comment already exists, use it.
+    if (comment) {
+        return comment;
+    }
+    
+    // Otherwise create a new Comment.
+    NSString *entityName = NSStringFromClass([Comment class]);
+    comment = [NSEntityDescription insertNewObjectForEntityForName:entityName inManagedObjectContext:self.managedObjectContext];
+    
+    [self updateComment:comment withRemoteComment:remoteComment];
+    comment.post = post;
+    comment.depth = depth;
+    comment.hierarchy = hierarchy;
```

Comment:
The hierarchy isn't calculated correctly with the Comment in isolation (i.e. without it's ancestors). It is calculated correctly when the full hierarchical comments are refreshed, but to get the comment to display properly when restored, the depth and hierarchy need to be provided.